Extend anonymisation settings to allow them to be site specific

michalkleiner opened the pull request:
### Description:

Reuse the anonymisation vue component from global settings within site settings to configure anonymisation options per site.

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

michalkleiner committed:
Extend anonymisation settings to allow them to be site specific

michalkleiner committed:
Mark api methods as internal

michalkleiner committed:
Only set idSite if truthy

michalkleiner committed:
Mark UI-only API methods as ignored for system tests

michalkleiner committed:
Update UI test screenshots

michalkleiner committed:
Use correct cache setting method

michalkleiner committed:
Remove controller methods and use API instead

michalkleiner committed:
Use local variable

michalkleiner committed:
Remove ignore annotation so the method is available via API

michalkleiner committed:
Use better descriptive name

michalkleiner committed:
Fix test

michalkleiner committed:
Merge branch '5.x-dev' into dev-19315

michalkleiner committed:
Tweaks from testing

michalkleiner committed:
Add more privacy config tests

michalkleiner committed:
Merge branch '5.x-dev' into dev-19315

michalkleiner committed:
Fix CS and minor comment tweaks

michalkleiner committed:
Add expected system test files for new PrivacyManager API method

michalkleiner committed:
Fix CS

michalkleiner committed:
Remove forgotten logging

innocraft-automation committed:
Build vue files

michalkleiner committed:
Undo comment change as now covered by PHPStan baseline exception

michalkleiner committed:
Fix CS

michalkleiner committed:
Temporarily update submodule reference

michalkleiner committed:
Merge branch '5.x-dev' into dev-19315

michalkleiner committed:
Temporarily update submodule reference

michalkleiner committed:
Merge branch '5.x-dev' into dev-19315

michalkleiner committed:
Merge branch '5.x-dev' into dev-19315

michalkleiner committed:
Use site-specific config in more places

michalkleiner committed:
Fix CS after 5.x-dev merge

michalkleiner committed:
Add class to cancel button

michalkleiner committed:
Add classes to individual field wrappers for easier targetting in tests

michalkleiner committed:
Add site information to the top of anonymisation settings

michalkleiner committed:
Add site-specific privacy settings UI test and screenshots

michalkleiner committed:
Merge branch '5.x-dev' into dev-19315

michalkleiner committed:
Remove option to set cookieless tracking per site

michalkleiner committed:
Remove console debugging

michalkleiner committed:
Fix type hints in class annotation

michalkleiner committed:
Build dist files

michalkleiner committed:
Merge branch '5.x-dev' into dev-19315

michalkleiner committed:
Use instance privacy config as cookieless tracking can't be site-specific

michalkleiner committed:
Fix CS

michalkleiner committed:
Update UI test screenshots from CI

michalkleiner committed:
Update PHPStan baseline

michalkleiner committed:
Merge branch '5.x-dev' into dev-19315

michalkleiner committed:
Further removal of cookieless tracking related changes

mneudert commented on plugins/CoreHome/Tracker/VisitRequestProcessor.php:
```diff
@@ -155,8 +155,9 @@ public function afterRequestProcessed(VisitProperties $visitProperties, Request
          * This event is primarily used by the **PrivacyManager** plugin to anonymize IP addresses.
          *
          * @param string &$ip The visitor's IP address.
+         * @param mixed $idSite The site ID we're tracking the visit for.
```
```suggestion
         * @param int $idSite The site ID we're tracking the visit for.
```

At this point we know we have a site,. And with the trace through the various request functions "getIdSiteIfExists -> getIdSite -> getIdSiteUnverified -> getRequestVar(int)" this should always be an int. Though I would keep the int cast until PHPStan tells us the cast is definitely unnecessary 🙈

michalkleiner replied:
Yep, changed to int

mneudert commented on plugins/PrivacyManager/tests/Integration/PrivacyManagerConfigTest.php:
```diff
@@ -126,8 +129,40 @@ public function testSetTrackerCacheContentShouldGetValuesFromConfig()
+
+    public function testDirectSettingGlobalConfigIsReflectedViaFallbackForSite()
+    {
+        $this->config->setIdSite(null);
```
That call does nothing.

Is it really intented to never be able to reset the config to "all sites only" by passing `null`? Or is this call obsolete and can be removed?

michalkleiner replied:
Updated to accept nulls again to reset the config.

mneudert commented on plugins/PrivacyManager/API.php:
```diff
@@ -234,29 +228,114 @@ private function formatAvailableColumnsToAnonymize($columns)
-            IPAnonymizer::deactivate();
+        if (is_numeric($idSiteSpecific)) {
+            $idSite = intval($idSiteSpecific);
+            Piwik::checkUserHasAdminAccess($idSiteSpecific);
```
There is still a  `checkUserHasSuperUserAccess` call right before 👀

michalkleiner replied:
Moved that to inside the condition so only one check is performed based on if it's global or site-specific. Admin check will be true for all super admins early on.

mneudert commented on plugins/PrivacyManager/PrivacyManager.php:
```diff
@@ -72,6 +73,8 @@ class PrivacyManager extends Plugin
     private $dntChecker = null;
     private $ipAnonymizer = null;
 
+    private $logger = null;
```
Looks unused 🤔  Was there anything planned for it?

michalkleiner replied:
Yep, a leftover from the tests debugging kerfuffle.

mneudert commented on plugins/PrivacyManager/PrivacyManager.php:
```diff
@@ -206,15 +211,18 @@ public function shouldDisablePostProcessing(&$shouldDisable, $request)
     {
         if ($view->requestConfig->getApiModuleToRequest() === 'Referrers' && !$view->requestConfig->idSubtable) {
-            $config = new Config();
+            $idSite = $view->requestConfig->getRequestParam('idsite');
```
I think this will default to `string`, and not be an int, so best to add an `else { (int) $idSite }` part to have one complaint less later.

michalkleiner replied:
Added casting in an else branch 👍

mneudert commented on plugins/SitesManager/vue/src/SiteFields/SiteFields.vue:
```diff
@@ -211,6 +232,9 @@ import {
 } from 'CorePluginsAdmin';
+import {
+  AnonymizeIp,
+} from 'PrivacyManager';
```
This will need som fallbacks.

It is possible to deactivate the `PrivacyManager` plugin. Before it should just hide all the settings, now it breaks the site management completely. Something like `const AnonymizeIp = useExternalPluginComponent('PrivacyManager', 'AnonymizeIp');` and then try "AnonymizeIp !== null" to detect if the plugin is available.

And if not we should hide the privacy button 😬

michalkleiner replied:
Nice find, that wouldn't cross my mind that privacy manager can be disabled. Adjusted the code accordingly and added a UI test as well.

michalkleiner committed:
Update type hint

michalkleiner committed:
Allow re-setting anonymisation config for all sites

michalkleiner committed:
Remove forgotten logger

michalkleiner committed:
Ensure we always have ?int site id

michalkleiner committed:
Cater for possibility of PrivacyManager plugin being disabled

michalkleiner committed:
Adjust permissions check within the API setter

michalkleiner committed:
Add site-specific suffix to field names (when available) to prevent overriding between sites

michalkleiner committed:
Add UI test for scenario with disabled PrivacyManager plugin

michalkleiner committed:
Fix CS

michalkleiner committed:
Merge branch '5.x-dev' into dev-19315

sgiehl commented on plugins/PrivacyManager/API.php:
```diff
@@ -234,29 +228,113 @@ private function formatAvailableColumnsToAnonymize($columns)
+        $forceCookielessTracking = false,
+        $randomizeConfigId = false,
+        $idSiteSpecific = null,
+        $useSiteSpecificSettings = false
```
It would be beneficial to directly add type hints for all or at least for the new parameters when working on any API method. In this case it should be quite easy, as a parameter type hinted with boolean will also accept e.g. `0` or `1` .
And for `$idSiteSpecific` you could use `?int`, which would make the `is_numeric` check obsolete.

michalkleiner replied:
Ok, that's fair. Since it's an internal API, I've added types to all the params and removed casting within the method where not needed anymore.

michalkleiner committed:
Merge branch '5.x-dev' into dev-19315

# Conflicts:
#	plugins/PrivacyManager/vue/dist/PrivacyManager.umd.min.js

michalkleiner committed:
Add type hints to API setter

michalkleiner committed:
Update UI test selectors after making field IDs unique

michalkleiner committed:
Update UI test screenshot from CI

michalkleiner committed:
Fix UI test selectors

michalkleiner committed:
Merge branch '5.x-dev' into dev-19315

michalkleiner committed:
Merge branch '5.x-dev' into dev-19315

sgiehl commented on plugins/PrivacyManager/API.php:
```diff
@@ -234,49 +228,133 @@ private function formatAvailableColumnsToAnonymize($columns)
+        bool $forceCookielessTracking = false,
+        bool $randomizeConfigId = false,
+        ?int $idSiteSpecific = null,
+        bool $useSiteSpecificSettings = false
```
Renaming the API parameters might break the ActivityLog as it is looking for the old parameter names.

michalkleiner replied:
Reverted the change to method and param names. Created a corresponding ActivityLog premium plugin PR at https://github.com/innocraft/plugin-ActivityLog/pull/205, which also needed a version bump here.

michalkleiner committed:
Revert changes to API method name and params for ActivityLog compatibility

michalkleiner committed:
Bump version so that ActivityLog can support the new API params

michalkleiner committed:
Merge branch '5.x-dev' into dev-19315

sgiehl reviewed (dismissed):
No further concerns from my side.

michalkleiner committed:
Merge branch '5.x-dev' into dev-19315

michalkleiner committed:
Merge branch '5.x-dev' into dev-19315

# Conflicts:
#	plugins/TagManager

michalkleiner committed:
Move site-specific anonymisation settings into the standard edit form

michalkleiner committed:
Update UI tests

michalkleiner committed:
Merge branch '5.x-dev' into dev-19315

michalkleiner committed:
Don't show anonymisation settings when creating a new site

michalkleiner committed:
Update UI test screenshots

michalkleiner committed:
Revert sites manager button layout

michalkleiner committed:
Revert "Update UI test screenshots"

This reverts commit fd0282c39aae0acebe6a677bfabc9ef6b42e871c.

michalkleiner committed:
Add link to system-level privacy settings

michalkleiner committed:
Update UI test screenshots

michalkleiner committed:
Update TagManager submodule

michalkleiner committed:
Use site specific id selector for UTC help text

michalkleiner committed:
Hide UTC selector in UI tests

michalkleiner committed:
Merge branch '5.x-dev' into dev-19315

michalkleiner committed:
Update UI tests

michalkleiner committed:
Update UI test screenshots from CI

michalkleiner committed:
Update TagManager submodule

michalkleiner committed:
Update TagManager submodule

michalkleiner committed:
Increase maximum bundle size

michalkleiner committed:
Adjust per-site anonymisation to only use a single save button

The mechanism always triggers two API posts for both site settings and anonymisation settings even when

michalkleiner committed:
Update UI test and test screenshots after using a single set of save/cancel buttons

michalkleiner committed:
Update UI test screenshots from CI

sgiehl commented on plugins/PrivacyManager/Config.php:
```diff
@@ -62,30 +74,48 @@ public function __get($name)
     }
 
-    private function prefix($optionName)
+    private function prefix(string $optionName, bool $addIdSite = true)
```
```suggestion
    private function prefix(string $optionName, bool $addIdSite = true): string
```

sgiehl commented on plugins/PrivacyManager/PrivacyManager.php:
```diff
@@ -712,9 +728,9 @@ private static function getReportDate($strPeriod, $strDate)
         // if range, only look at the first date
         if ($strPeriod === 'range') {
-            $idSite = Common::getRequestVar('idSite', '');
+            $idSite = Common::getRequestVar('idSite', '0', 'int');
```
We should try to get rid of using `Common::getRequestVar` when working on things. Use `Piwik\Request::fromRequest()->getIntegerParameter('idSite', 0)` instead

michalkleiner committed:
Only trigger privacy settings save when editing existing site

sgiehl commented on plugins/PrivacyManager/lang/en.json:
```diff
@@ -182,6 +182,7 @@
         "AnonymizeOrderIdNote": "Because an Order ID can be cross-referenced with another system, typically an ecommerce shop, the Order ID may count as personal information under the GDPR. When you enable this option, an Order ID will be automatically anonymized so no personal information will be tracked.",
         "UseAnonymizeIp": "Anonymize Visitors' IP addresses",
         "UseAnonymizeTrackingData": "Anonymize Tracking Data",
+        "UseAnonymizeTrackingDataForThisSite": "Tracking data anonymization settings for this site",
```
Should it's be clear that the settings are for the specific site, now that they are displayed within the according site settings?
```suggestion
        "UseAnonymizeTrackingDataForThisSite": "Tracking data anonymization settings",
```

sgiehl commented on plugins/PrivacyManager/tests/UI/PrivacyManager_SiteSpecific_spec.js:
```diff
@@ -0,0 +1,223 @@
+        await page.waitForNetworkIdle();
+        await hideUTCTimeInfo();
+
+        await capturePage('save_site_specific_settings_site1');
```
I guess we could actually avoid taking screenshots for such tests. It might be enough to check that the save notification appears. The "collapsing" site settings should already be covered by SitesManager tests

michalkleiner replied:
Fair enough, I'll keep that in mind going forward. If it's not a major blocker keeping it as is for now that would be great.

sgiehl commented on plugins/PrivacyManager/Config.php:
```diff
@@ -62,30 +74,48 @@ public function __get($name)
     }
 
-    private function getFromOption($name, $config)
+    private function getFromTrackerCache($name, $config)
```
```suggestion
    private function getFromTrackerCache(string $name, array $config)
```

sgiehl commented on plugins/SitesManager/vue/src/SiteFields/SiteFields.vue:
```diff
@@ -160,6 +154,22 @@
           </div>
         </div>
 
+        <template v-if="privacyManagerEnabled && theSite && theSite.idsite">
```
From a usability perspective it's a bit weird, that you can only set the privacy settings, once the page was already created. There is no way to directly provide them when the site is created.

michalkleiner replied:
Yeah, when it was in a separate tab it was easier as you wouldn't get to that view until saved. I'll see if it's easy to get some defaults in and only set when relevant, but this is already taking sooo long that I'd prefer to not do any more functional changes if at all possible.

michalkleiner replied:
With the removal of the config id feature flag and having to confirm that specific setting with a password to a separate API endpoint, I think without larger refactoring to a new single endpoint that would cover both site settings and privacy settings (currently 2 plugins) and both add/edit actions for the websites, displaying the privacy settings (and then correctly saving them at the right time) when creating a site is, in my opinion, very difficult.

sgiehl commented on plugins/PrivacyManager/Config.php:
```diff
@@ -62,30 +74,48 @@ public function __get($name)
+        $valueSite = $this->getFromSpecificTrackerCache($name, $cache, $config, $useFallback = false);
+        $valueGeneralWithFallback = $this->getFromSpecificTrackerCache($name, $generalCache, $config);
+
+        return (null !== $valueSite ? $valueSite : $valueGeneralWithFallback);
```
Wouldn't that be the same
```suggestion
        return $valueSite ?? $valueGeneralWithFallback;
```

sgiehl commented on plugins/PrivacyManager/API.php:
```diff
@@ -235,51 +229,126 @@ private function formatAvailableColumnsToAnonymize($columns)
 
     /**
+     * Provide tracker file name and whether it's writable
+     */
```
We should try to add some more specific type hinting in the doc blocks where possible
```suggestion
     * @return array{0: string, 1: bool}
     */
```

sgiehl commented on plugins/PrivacyManager/DoNotTrackHeaderChecker.php:
```diff
@@ -35,13 +35,14 @@ public function __construct($config = null)
             return;
         }
 
+        $this->config->setIdSite((int) $request->getIdSiteIfExists());
```
That setting is available only globally, so guess that wouldn't be needed. On the other side it won't hurt if we ever decide to make it site specific...

michalkleiner replied:
I'd keep it, we still work with an instance of the config, so the overhead is minimal.

sgiehl reviewed (dismissed):
Didn't have time to do a full review. Left some comments and questions that came up when looking into the code.

In addition I was wondering if we should write some more integration or system tests around that, which ensure that the anonymization settings are correctly respected.
There should already be tests that turn the global anonymization one, but we could extend them to also have different settings per site and ensure that the values are correctly tracked for all cases then.

michalkleiner committed:
Merge branch '5.x-dev' into dev-19315

michalkleiner committed:
Apply review feedback

michalkleiner committed:
Update UI test screenshots from CI

michalkleiner committed:
Refactor fixture to prepare tracker in a reusable way

michalkleiner committed:
Update type hint to accept floats when adding hours

michalkleiner committed:
Track extra visit to side id 2

michalkleiner committed:
Add test for IP anonymisation settings

michalkleiner committed:
Fix getting global fallback value before resolving to default value

michalkleiner committed:
Add expected API response files for site-specific anonymisation

The files should be identical to their counterpart without the site2 in their name, apart from id site, id visit, IP addresses and fingerprints.

michalkleiner committed:
Merge branch '5.x-dev' into dev-19315

Left out some changes that need to be reimplemented manually and will be in a follow up commit.

michalkleiner committed:
Update implementation after 5.x-dev merge

# Conflicts:
#	plugins/PrivacyManager/Controller.php

michalkleiner committed:
Ensure password is required when setting config id randomisation per site

michalkleiner committed:
Merge branch '5.x-dev' into dev-19315

michalkleiner committed:
Remove config id randomisation feature flag from API response for settings and update tests

michalkleiner committed:
Add tests for config id randomisation using per-site settings

michalkleiner committed:
Remove unused method

michalkleiner committed:
Add type hints to API setter

Re-add typehints to API params

michalkleiner committed:
Fix variable name after refactoring

michalkleiner committed:
Fix CS

michalkleiner committed:
Fix API param name after refactoring

michalkleiner committed:
Update UI test screenshots

michalkleiner committed:
Merge branch '5.x-dev' into dev-19315

michalkleiner committed:
Update UI test screenshots

michalkleiner committed:
Ensure password is only required when enabling config id randomisation

Previously the password would be required even when it was enabled before per site and using per-site settings was disabled, the value would still be reflected.
With this change, it is only reflected when it's visible, i.e. in system settings context or when per-site settings are enabled (visible).

michalkleiner committed:
Disable saving state when success notification is displayed

caddoo committed:
Merge branch '5.x-dev' into dev-19315

caddoo committed:
Merge branch '5.x-dev' into dev-19315

michalkleiner committed:
Merge branch '5.x-dev' into dev-19315

michalkleiner committed:
Merge branch '5.x-dev' into dev-19315

michalkleiner committed:
Update config to work with compliance policy mechanism and per-site settings

michalkleiner committed:
Improve code legibility and add docblocks

michalkleiner committed:
Rebuild dist file

michalkleiner committed:
Use Config's mechanism to get raw option value

michalkleiner committed:
Fix type

michalkleiner committed:
Fix namespace

michalkleiner committed:
Update test to cater for ip mask length and ip anonymisation being set per site

michalkleiner committed:
Revert baseline change

michalkleiner commented:
I don't think the TagManagerTeaser UI test failure is related to changes here.

nathangavin commented on plugins/PrivacyManager/Settings/IPAnonymisation.php:
```diff
@@ -39,9 +39,15 @@ public function getValue()
+        return (new Config($idSite))->prefix('ipAnonymizerEnabled');
+    }
+
+    public static function getOptionValue(?int $idSite = null): ?string
```
This function is unnecessary, updating the `getOptionName()` should be sufficient. We shouldn't worry about the handling of a default global value from within these setting classes.

nathangavin commented on plugins/PrivacyManager/Settings/IpAddressMaskLength.php:
```diff
@@ -39,9 +39,15 @@ public function getValue()
+        return (new Config($idSite))->prefix('ipAddressMaskLength');
+    }
+
+    public static function getOptionValue(?int $idSite = null): ?string
```
Same as the `IPAnonymisation` setting, this function is unnecessary.

nathangavin reviewed (dismissed):
The CNIL stuff looks pretty good, but I think there is a little too much introduced complexity into the settings classes, they are meant to only retrieve db data and compare to policy, so I wouldn't worry about handling backup global values from inside the settings classes.

michalkleiner committed:
Remove getOptionValue overrides

michalkleiner committed:
Revert "Remove getOptionValue overrides"

This reverts commit 473b0f611bc7fb174a2b3b74b5a499ec3464f98a.

michalkleiner committed:
Add a custom setting interface for cases where the setting value is provided in some more complex way

michalkleiner committed:
Use custom setting interface

nathangavin commented:
> The CNIL stuff looks pretty good, but I think there is a little too much introduced complexity into the settings classes, they are meant to only retrieve db data and compare to policy, so I wouldn't worry about handling backup global values from inside the settings classes.

Reviewed and discussed with @michalkleiner, I am happy with the CNIL-related changes.

michalkleiner committed:
Merge branch '5.x-dev' into dev-19315

# Conflicts:
#	plugins/TagManager

mneudert commented on core/Settings/Interfaces/CustomSettingInterface.php:
```diff
@@ -0,0 +1,8 @@
+<?php
+
+namespace Piwik\Settings\Interfaces;
+
+interface CustomSettingInterface
+{
+    public static function getCustomValue(?int $idSite = null);
+}
```
Entirely optional until we bump our PHPStan level to 9+ , but would match the other interfaces:

```suggestion
/**
 * @template T of mixed
 */
interface CustomSettingInterface
{
    /**
     * @return T
     */
    public static function getCustomValue(?int $idSite = null);
}
```

And the matching `@implements CustomSettingInterface<whatever-type-is-correct>` in the places using it.

michalkleiner replied:
Added

mneudert commented on plugins/PrivacyManager/Config.php:
```diff
@@ -67,43 +81,93 @@ public function __get($name)
     {
-        return 'PrivacyManager.' . $optionName;
+        // if requested, adding the site ID in the middle to have all the site-specific settings together
+        return 'PrivacyManager.' . (($addIdSite && $this->idSite) ? "idSite($this->idSite)." : '') . $optionName;
```
We should delete those new site specific options when a site is deleted.

The event (see e.g. [Goals plugin](https://github.com/matomo-org/matomo/blob/be81d28b79efc6795c07aa14535e30a9b6b1df7a/plugins/Goals/Goals.php#L105)) os readily available and we can prevent our options able from bloating more than necessary.

michalkleiner replied:
Good suggestion, implemented.

mneudert commented on tests/PHPUnit/Integration/ReleaseCheckListTest.php:
```diff
@@ -621,7 +621,7 @@ public function testTotalPiwikFilesSizeIsWithinReasonnableSize()
             $this->markTestSkipped("Skipped this test on local dev environment.");
         }
-        $maximumTotalFilesizesExpectedInMb = 62;
+        $maximumTotalFilesizesExpectedInMb = 63;
```
_Using a random line to get a resolvable thread._

The TagManager branch needs an update (and a new submodule update here). By now the linked branch is outdated and merging the update would revert several changes until the submodule gets updated again. Would also solve some confusion around the failing UI test that is only indirectly related to the outdated branch.

michalkleiner replied:
TagManager branch rebased, submodule updated here, thanks for pointing that out.

michalkleiner committed:
Update TagManager submodule

michalkleiner committed:
Add interface types

michalkleiner committed:
Add explicit API method param type

michalkleiner committed:
Ensure site-specific settings are removed when a site is deleted

michalkleiner committed:
Merge branch '5.x-dev' into dev-19315

michalkleiner committed:
Update expected system tests to a correct global default value

michalkleiner committed:
Update UI test screenshot from CI
